src/hotspot/share/gc/shenandoah/shenandoah_globals.hpp

```diff
@@ -35,6 +35,28 @@
                             range,                                          \
                             constraint)                                     \
                                                                             \
+  product(bool, GenShenCensusAtEvac, true, EXPERIMENTAL,                    \
+          "Object age census at evacuation, not mearking")                  \
+                                                                            \
+  product(bool, GenShenAdaptiveTenuring, false, EXPERIMENTAL,               \
+          "Adaptive tenuring threshold")                                    \
+                                                                            \
+  product(uintx, GenShenMinTenuringThreshold, 1, EXPERIMENTAL,              \
+          "Floor for adaptive tenuring threshold")                          \
+                                                                            \
+  product(uintx, GenShenMaxTenuringThreshold, 16, EXPERIMENTAL,             \
```

Comment:
Why not 15?

Reply:
We'll need to discuss this. Allowing this to be 16 (i.e. 1 greater than max_age) allows for a potential "never tenure" behavior. Otherwise, age 15 objects would always become eligible for tenuring. I think allowing it to be above 15 (i.e. allowing 16, and allowing objects to stay in young till and including age 15) would seem to make sense.

I'll set the default to 15 (thus never allowing objects that , but allow users to set it to 16, simulating "never tenure", if both min and max are set to 16.

Reply:
Provide guidance here on how to use these min and max values to obtain different behavior, while picking sensible defaults (i.e. 0 and 15).

Reply:
Done. Updated guidance in documentation.